examples_tests/0.ImportanceSamplingEnvMaps/main.cpp

```diff
@@ -0,0 +1,668 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#define _NBL_STATIC_LIB_
+#include <nabla.h>
+#include "../../../include/nbl/asset/filters/CSummedAreaTableImageFilter.h"
+#include "nbl/ext/FullScreenTriangle/FullScreenTriangle.h"
+#include "nbl/ext/ScreenShot/ScreenShot.h"
+#include "glm/glm.hpp"
+#include "../common/QToQuitEventReceiver.h"
+
+using namespace nbl;
+using namespace asset;
+using namespace core;
+using namespace video;
+
+using SATFilter = CSummedAreaTableImageFilter<false>;
+
+static core::smart_refctd_ptr<ICPUBuffer> computeLuminancePdf(smart_refctd_ptr<ICPUImage> envmap)
+{
+	const core::vector2d<uint32_t> envmapExtent = { envmap->getCreationParameters().extent.width, envmap->getCreationParameters().extent.height };
+	const uint32_t channelCount = getFormatChannelCount(envmap->getCreationParameters().format);
+	
+	const core::vector2d<uint32_t> pdfDomainExtent = { envmapExtent.X, envmapExtent.Y };
+
+	const size_t outBufferSize = pdfDomainExtent.X * pdfDomainExtent.Y * sizeof(double);
+	core::smart_refctd_ptr<ICPUBuffer> outBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(outBufferSize);
+
+	const double luminanceScales[4] = { 0.2126729 , 0.7151522, 0.0721750, 0.0 };
+
+	float* envmapPixel = (float*)envmap->getBuffer()->getPointer();
+	double* outPixel = (double*)outBuffer->getPointer();
+
+	for (uint32_t y = 0; y < pdfDomainExtent.Y; ++y)
+	{
+		const double sinTheta = core::sin(core::PI<double>() * ((y + 0.5) / (double)pdfDomainExtent.Y));
+
+		for (uint32_t x = 0; x < pdfDomainExtent.X; ++x)
+		{
+			double result = 0.0;
+			for (uint32_t ch = 0; ch < channelCount; ++ch)
+				result += luminanceScales[ch] * envmapPixel[ch];
+
+			*outPixel++ = result * sinTheta;
+			envmapPixel += channelCount;
+		}
+	}
```

Comment:
the buffer's pointer being the first pixel, wont always work like that, there are offsets, strides, formats (you can't just cast to float), etc.